File: week2/java/week1_10.java
import java.util.ArrayList;

public class week1_10 {
    public static int solution(int bridge_length, int weight, int[] truck_weights){
        int totalTime = 1;
        int index = 1;
        int finished = 0;

        int[] startTime = new int[truck_weights.length];
        ArrayList<Integer> nowq = new ArrayList<>();
        nowq.add(0);
        startTime[0] = 1;

        while(finished < truck_weights.length){
            totalTime++;
            int now_weights = 0;

            for(int i=0; i<nowq.size(); i++){
                if(bridge_length - (totalTime - startTime[nowq.get(i)]) == 0){
                    nowq.remove(i);
                    finished += 1;
                }
            }

            if(index < truck_weights.length){
                for(int i=0; i<nowq.size(); i++) {
                    now_weights += truck_weights[nowq.get(i)];
                }
                if(now_weights+truck_weights[index] <= weight){
                    nowq.add(index);
                    startTime[index] = totalTime;
                    index++;
                }
            }


        }
        return totalTime;
    }
    public static void main(String[] args){
        int bridge_length = 2;
        int weight = 10;
        int[] truck = {7,4,5,6};

        System.out.println(solution(bridge_length, weight, truck));
    }
}

File: week1/c++/week1_10.cpp
#include <string>
#include <vector>
#include <queue>
#include <iostream>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights)
{
    int time = 1;
    vector<int> startTime(truck_weights.size(), 0);

    int count = 0;
    int index = 0;

    queue<int> q;
    q.push(index);
    startTime[index] = time;
    index++;

    while (count < truck_weights.size())
    {
        time++;
        int now = 0;

        // 트럭이 다리를 다 건넜는지 확인
        for (int i = 0; i < q.size(); i++)
        {
            if (bridge_length - (time - startTime[q.front()]) == 0)
            {
                q.pop();
                count++;
            }
        }

        // 다리가 버틸 수 있는 중량인지 확인
        if (index < truck_weights.size())
        {
            queue<int> temp = q;
            while (!temp.empty())
            {
                now += truck_weights[temp.front()];
                temp.pop();
            }

            // 새 트럭을 추가할 수 있는지 확인
            if (now + truck_weights[index] <= weight)
            {
                q.push(index);
                startTime[index] = time;
                index++;
            }
        }
    }
    return time;
}